packages/inter-protocol/src/auction/auctioneer.js

```diff
@@ -0,0 +1,284 @@
+import '@agoric/zoe/exported.js';
+import '@agoric/zoe/src/contracts/exported.js';
+import '@agoric/governance/exported.js';
+
+import { Far } from '@endo/marshal';
+import { E } from '@endo/eventual-send';
+import {
+  M,
+  makeScalarBigMapStore,
+  provideDurableMapStore,
+} from '@agoric/vat-data';
+import { AmountMath } from '@agoric/ertp';
+import {
+  atomicRearrange,
+  makeRatioFromAmounts,
+  makeRatio,
+  natSafeMath,
+  floorMultiplyBy,
+  provideEmptySeat,
+} from '@agoric/zoe/src/contractSupport/index.js';
+import { handleParamGovernance } from '@agoric/governance';
+import { makeTracer } from '@agoric/internal';
+import { FullProposalShape } from '@agoric/zoe/src/typeGuards.js';
+
+import { makeAuctionBook } from './auctionBook.js';
+import { BASIS_POINTS } from './util.js';
+import { makeScheduler } from './scheduler.js';
+import { auctioneerParamTypes } from './params.js';
+
+/** @typedef {import('@agoric/vat-data').Baggage} Baggage */
+
+const { Fail, quote: q } = assert;
+
+const trace = makeTracer('Auction', false);
+
+const makeBPRatio = (rate, currencyBrand, collateralBrand = currencyBrand) =>
+  makeRatioFromAmounts(
+    AmountMath.make(currencyBrand, rate),
+    AmountMath.make(collateralBrand, 10000n),
+  );
+
+/**
+ * @param {ZCF<GovernanceTerms<{
+ *   StartFrequency: 'relativeTime',
+ *   ClockStep: 'relativeTime',
+ *   StartingRate: 'nat',
+ *   lowestRate: 'nat',
+ *   DiscountStep: 'nat',
+ * }> & {
+ *   timerService: import('@agoric/time/src/types').TimerService,
+ *   priceAuthority: PriceAuthority
+ * }>} zcf
+ * @param {{
+ *   initialPoserInvitation: Invitation,
+ *   storageNode: StorageNode,
+ *   marshaller: Marshaller
+ * }} privateArgs
+ * @param {Baggage} baggage
+ */
+export const start = async (zcf, privateArgs, baggage) => {
+  const { brands, timerService: timer, priceAuthority } = zcf.getTerms();
+  timer || Fail`Timer must be in Auctioneer terms`;
+  const timerBrand = await E(timer).getTimerBrand();
+
+  const books = provideDurableMapStore(baggage, 'auctionBooks');
+  const deposits = provideDurableMapStore(baggage, 'deposits');
+  const brandToKeyword = provideDurableMapStore(baggage, 'brandToKeyword');
+
+  const reserveFunds = provideEmptySeat(zcf, baggage, 'collateral');
+
+  const addDeposit = (seat, amount) => {
+    const depositListForBrand = deposits.get(amount.brand);
+    deposits.set(
+      amount.brand,
+      harden([...depositListForBrand, { seat, amount }]),
+    );
+  };
+
+  // Called "discount" rate even though it can be above or below 100%.
+  /** @type {NatValue} */
+  let currentDiscountRateBP;
+
+  const distributeProceeds = () => {
+    for (const brand of deposits.keys()) {
+      const book = books.get(brand);
+      const { collateralSeat, currencySeat } = book.getSeats();
+
+      const depositsForBrand = deposits.get(brand);
```

Comment:
if `depositsForBrand.length === 0` this function does nothing. is that intended or an unexpected state?